parlai/agents/bert_ranker/crossencoder_ranker.py

```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+from parlai.core.torch_ranker_agent import TorchRankerAgent
+from .bert_dictionary import BertDictionaryAgent
+from .helpers import get_bert_optimizer, BertWrapper, BertModel, \
+    add_common_args, surround
+import torch
+import json
+
+
+class CrossEncoderRankerAgent(TorchRankerAgent):
+    """ TorchRankerAgent implementation of the crossencoder.
+        It is a standalone Agent. It might be called by the Both Encoder.
+    """
+
+    @staticmethod
+    def add_cmdline_args(parser):
+        add_common_args(parser)
+
+    def __init__(self, opt, shared=None):
+        opt['rank_candidates'] = True
+        opt["bert_id"] = 'bert-base-uncased'
+        super().__init__(opt, shared)
+        # NOTE: This is done AFTER init so that it's after load(), on purpose.
+        # the state dict of MyModule and DataParallel(MyModule) is not the same
+        if self.opt["multigpu"]:
+            self.model = torch.nn.DataParallel(self.model)
+        self.clip = -1
+        self.NULL_IDX = self.dict.pad_idx
+        self.START_IDX = self.dict.start_idx
+        self.END_IDX = self.dict.end_idx
+
+    def build_model(self):
+        self.model = BertWrapper(
+            BertModel.from_pretrained(
+                self.opt["bert_id"]),
+            1,
+            add_transformer_layer=self.opt["add_transformer_layer"],
+            layer_pulled=self.opt["pull_from_layer"])
+
+    def init_optim(self, params, optim_states=None, saved_optim_type=None):
+        if all(f in self.opt for f in ["num_samples", "num_epochs", "batchsize"]):
```

Comment:
Add a comment explaining this please. It's quite confusing?

Reply:
Actually it's something well worth a chat:
  the scheduler recommended by Devlon et al. decreases linearly. So I need to get the number of iterations (which can be computed as num_samples * num_epoch / size_batch).

However, I couldn't get the number of samples from the task. So for now I added it as a parameter. Any idea?

Reply:
So i just added invsqrt LR scheduling, and followed the same approximation as fairseq: number of calls to update_params()